### DISCUSSION TITLE
Calling Internal Route(Function) with Depends as Parameter

### DISCUSSION THREAD

**Discussion by vijaystroup** [fastapi/fastapi#6993] [Questions] [answered]
Python: 3.8.5
FastAPI: 0.63.0

Before, I had a route defined as something along the lines of:
```python
@router.get('/customers')
def get_customers(company: str, name: str):
    """get array of customers for employee"""

    ...
```
With this, I would call this route from another route like the following:
```python
bookings = get_customers(company, name)
```

Recently I changed my security protocols, so now my customers route is defined as follows:
```python
@router.get('/customers')
def get_customers(user=Depends(auth.verify)):
    """get array of customers for employee"""

    ...
```

Is there a way (without creating a request from python), to call this function like I was before with a different parameter or would it be better to take some of the logic from within the customers route and put it in a helper function that both can use?

Here is my Auth Class that is being Depended:
```python
class Auth:
    security = HTTPBearer()
    secret = secrets['JWT_SIGNATURE']

    def __init__(self, mongo):
        self.mongo = mongo

    def encode_token(self, _id: str, db: str):
        """return encoded jwt token after logging in"""

        return jwt.encode({
            'exp': datetime.utcnow() + timedelta(minutes=10),
            'iat': datetime.utcnow(),
            '_id': _id,
            'db': db
            },
            self.secret,
            algorithm='HS256'
        )

    def decode_token(self, token):
        """decode jwt token and return user from db if valid"""

        try:
            payload = jwt.decode(token, self.secret, algorithms=['HS256'])
            user = self.mongo[payload['db']]['employees'].find_one(
                {'_id': ObjectId(payload['_id'])}
            )
            if not user:
                raise HTTPException(status_code=404, detail='User not found')
            company = self.mongo['businesses']['entities'].find_one({'db': payload['db']})
        except jwt.ExpiredSignatureError:
            raise HTTPException(status_code=401, detail='Signature has expired')
        except jwt.InvalidTokenError as e:
            raise HTTPException(status_code=401, detail='Invalid token')

        user['company'] = company
        user['db'] = payload['db']
        return user

    def verify(self, auth: HTTPAuthorizationCredentials = Security(security)):
        """wrapper function for routes"""

        return self.decode_token(auth.credentials)
```

**Comment by falkben** [comment, marked as answer]
I would personally go with the approach of refactoring the logic out of your `get_customers` method and trying to reuse it, assuming it should actually be reused in both an authenticated and unauthenticated endpoint.

If a user isn't authenticated, you presumably don't want them accessing that method anyways.

If you'd rather redirect to that endpoint, to get them to go through the auth flow, you can do that too with a RedirectResponse. From Starlette there's: `app.url_path_for()` so you don't have to hard code the url. See their docs: https://www.starlette.io/routing/#reverse-url-lookups

**Comment by vijaystroup** [comment]
> I would personally go with the approach of refactoring the logic out of your `get_customers` method and trying to reuse it, assuming it should actually be reused in both an authenticated and unauthenticated endpoint.
> 
> If a user isn't authenticated, you presumably don't want them accessing that method anyways.
> 
> If you'd rather redirect to that endpoint, to get them to go through the auth flow, you can do that too with a RedirectResponse. From Starlette there's: `app.url_path_for()` so you don't have to hard code the url. See their docs: https://www.starlette.io/routing/#reverse-url-lookups

Thanks for the response @falkben !

Before my endpoints were all "unauthenticated" but I had logic within the function to check to see if they were who they said they were. There is only a very small piece of logic in my `get_customers` function that can be dispatched to a helper function that `get_customers` and my new authenticated route can both call.

I was just curious to see if there was a way to call it like a normal function when depending on fastapi security. I'll leave this issue open for a little longer if anyone else want's to add anything but as of now, breaking out the logic to a helper function is the method I'm thinking of right now.

**Comment by tiangolo** [comment]
Thanks for the help here @falkben ! 👏  🙇 

Thanks for reporting back and closing the issue @VijayStroup  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
